rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,364 @@
+# Feature Name: Generic UI Navigation API
+
+## Summary
+
+Introduce [`bevy-ui-navigation`] into the bevy tree.
+
+By default this, amounts to replacing `Interaction` with `Focusable` in `bevy_ui`.
+On top of the current behavior, this adds the following capabilities:
+- Focus management for **gamepads**
+- Decouples input from focus management
+- Decouples focus management from `bevy_ui`
+- User-specified input handling for UI interaction
+- An optional menu tree management system
+
+## Terminology
+
+- _menu tree_: the access hierarchy of menus.
+- _activating_ a focusable: Sending the `Action` event while this focusable is focused.
+- _focused menu_: The menu in which the currently focused `Focusable` is.
+- _child menu_ (of a focusable): The menu accessed from this focusable.
+
+## Motivation
+
+`bevy_ui` interaction story is crude.
+The only way to interact with the UI currently
+is through the `Interaction` and `FocusPolicy` components.
+`Interaction` only supports mouse or touch-based input.
+We want gamepad supports and more game-oriented functionalities,
+such as menu trees and movement between menus.
+
+The current UI focus system is limited to `bevy_ui` and
+can't be swapped out for something the user deems more appropriate.
+
+This is in contradiction with the general philosophy of bevy,
+which is of providing to the user all the tools necessary to create something
+that fits their needs optimally.
+
+This RFC proposes to resolve all those questions.
+[`bevy-ui-navigation`] provides a highly customizable ECS-based navigation engine.
+
+## User-facing explanation
+
+The [bevy-ui-navigation README][`bevy-ui-navigation`] is a good start.
+To summarize, here are the headlines:
+- The navigation system can only be interacted with with events:
+  `NavRequest` as input and `NavEvent` as output.
+- `NavEvent::FocusChanged` contains the list of entities that are traversed
+  to go from the last focus position to the new one.
+- ui-navigation provides default input systems, 
+  but it's possible to disable them and instead use custom ones.
+  This is why we restrict interactions with navigation to events.
+- ui-navigation provides a system to declare buttons as "leading to menu X"
+  in order to create menu trees.
+- ui-navigation defines [custom `SystemParam`][event_helpers]s
+  to ease reaction to events (such a button press or menu change).
+  The most naive usage results into [elm architecture][elm_architecture]-like code.
+  Systems updating the model (ECS hierarchy) and systems reacting to events. 
+- It is possible to create isolated menus by using the `NavMenu` component.
+- All `Focusable` children in the `NavMenu` entity's tree will be part
+  of this menu, and moving using a gamepad from one focusable in a menu
+  can only lead to another focusable within the same menu.
+  You can specify a focusable in any other menu that will lead to this
+  one once activated.
+
+## Implementation
```

Comment:
The other edge case I want examined is "multiple cursors", like in Smash Bros character select. This was raised in Discord and it seemed like you had a good fix, but it's not in this RFC.

Reply:
This might be better reserved for a "future possibilities" section. This RFC already skips a few non-necessary/nice-to-have features from `bevy-ui-navigation`, and it's still massive.

Reply:
Agreed, I'm on board with that.